### PR TITLE
Several Switchover fixes

### PR DESCRIPTION
**Pull request by tiagolobocastro**
test(switchover/bdd): add robustness tests
    
    Add tests to help exercise some corner cases seen during bug study.
    Tbh these are not very precise as it's difficult to re-create some corner cases,
    but nonetheless they are probably a decent starting point and can be used to help
    manually try to recreate these issues in a more automated fashion.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    fix(switchover): adds several corner case fixes
    
    Replace path is not idempotent in case the connection was established without moving path.
    gRPC Timeouts for connection are not enough which leads into moving the target
    when not necessary.
    Report of failed paths does not differentiate errors at all.
    Report of failed paths returns only 1 error.. we should have an error per path as we report.
     in batches.
    In case of no resources for republish volume the ha cluster retries in a loop with no delay.
    or backoff..
    In error case we don’t clean up shutdown targets.
    On switchover failure, failed paths are not cleared in the cluster agent.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    chore(python/bdd): pin to version 6
    
    Newer version (not sure which exactly) seems to generate the test cases without a name.
    todo: determine best version to use
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    fix(ha-node): add subsystem sync for platform none
    
    On the deployer the udev code we have for listening to disk events does not seem to work.
    For now, assume it is incompatible and add a subsystem sync just for platform none.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

### CODE REVIEW

File: control-plane/agents/src/bin/ha/cluster/switchover.rs [hunk 1]
```diff
@@ -49,12 +48,33 @@ pub enum Stage {
...
+        *self.0.read()
+    }
+}
+
+impl PartialEq for SwitchOverStage {
+    fn eq(&self, other: &Self) -> bool {
+        *self.0.read() == *other.0.read()
+    }
+}
+impl Eq for SwitchOverStage {}
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/ha/cluster/switchover.rs:70]: Is this supposed to be empty?

File: control-plane/agents/src/bin/ha/node/path_provider.rs [hunk 2]
```diff
@@ -92,9 +97,41 @@ impl CachedNvmePathProvider {
     pub fn start(&mut self) -> anyhow::Result<impl std::future::Future<Output = ()> + '_> {
         Ok(ready(()))
     }
+    #[cfg(target_os = "linux")]
+    fn spawn_nvmeadm_syncer(&self) {
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/ha/node/path_provider.rs:101]: 👍🏻

File: control-plane/stor-port/src/transport_api/mod.rs [hunk 3]
```diff
@@ -494,10 +516,14 @@ impl RequestMinTimeout {
     pub fn pool(&self) -> Duration {
         self.pool
     }
-    /// minimum timeout for nexus shutdown
+    /// minimum timeout for nexus shutdown.
```
Comment (Abhinandan-Purkait) [inline control-plane/stor-port/src/transport_api/mod.rs]: Start with capital? Below as well?

File: tests/bdd/common/nvme.py [hunk 4]
```diff
@@ -69,6 +69,7 @@ def nvme_connect(uri):
     hostnqn = nvme_hostnqn_arg(u)
 
     command = f"sudo nvme connect -t tcp -s {port} -a {host} -n {nqn} {hostnqn}"
+    print(command)
```
Comment (Abhinandan-Purkait) [inline tests/bdd/common/nvme.py:72]: Do we need this print?

File: tests/bdd/features/ha/robustness.feature [hunk 5]
```diff
@@ -0,0 +1,30 @@
+Feature: Switchover Robustness
+
+  Background:
+    Given a deployer cluster
+
+  Scenario: reconnecting the new target times out
+    Given a single replica volume
+    And a connected nvme initiator
+    And a reconnect_delay set to 15s
+    When we restart the volume target
```
Comment (Abhinandan-Purkait) [inline tests/bdd/features/ha/robustness.feature]: `When we restart the volume target node` ?

File: tests/bdd/features/ha/robustness.feature [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    And the ha clustering republishes
+    Then the path should be established
+
+  Scenario: path failure with no free nodes
+    Given a 2 replica volume
+    And a connected nvme initiator
+    When we cordon the non-target node
+    And we stop the volume target
```
Comment (Abhinandan-Purkait) [inline tests/bdd/features/ha/robustness.feature]: `And we stop the volume target node` ? Below as well?

File: tests/bdd/requirements.txt [hunk 7]
```diff
@@ -1,4 +1,4 @@
-pytest-bdd
+pytest-bdd==6.0.0
```
Comment (Abhinandan-Purkait) [inline tests/bdd/requirements.txt:1]: Should we revert this as this `def _():` seems to be the way forward?
As per - https://github.com/pytest-dev/pytest-bdd/pull/536

**Review by Abhinandan-Purkait — APPROVED** [review]
LGTM. Few nits.

File: control-plane/agents/src/bin/ha/cluster/switchover.rs (hunk 1 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs:70]: Yes, it's simply to indicate equivalence.

File: control-plane/stor-port/src/transport_api/mod.rs (hunk 3 above)
Comment (tiagolobocastro) [reply control-plane/stor-port/src/transport_api/mod.rs]: Modified all of them.

File: tests/bdd/common/nvme.py (hunk 4 above)
Comment (tiagolobocastro) [reply tests/bdd/common/nvme.py:72]: It's just a helper in case things fail, nothing gets printed when running pytest with default args.

File: tests/bdd/requirements.txt (hunk 7 above)
Comment (tiagolobocastro) [reply tests/bdd/requirements.txt:1]: Let's keep it as is for now, and let this soak, just in case it gets changed upstream..
Then to remove this we should change all our testcode to def _()

File: tests/bdd/features/ha/robustness.feature (hunk 6 above)
Comment (tiagolobocastro) [reply tests/bdd/features/ha/robustness.feature]: Done to all

**Comment by tiagolobocastro** [comment]
bors merge

**Review by abhilashshetty04 — APPROVED** [review]

**Comment by tiagolobocastro** [comment]
bors merge
